### PR TITLE
Remove code which updates `application_services_state.last_txn`

### PR DESCRIPTION
**Pull request by richvdh**
This column is unused as of #12209, so let's stop writing to it.

Fixes: #12352

### CODE REVIEW

**Review by erikjohnston — COMMENTED** [review]
Seems sensible, but there is a failing test

File: synapse/storage/databases/main/appservice.py [hunk 1]
```diff
@@ -452,16 +454,15 @@ async def set_appservice_stream_type_pos(
...
+        # this may be the first time that we're recording any state for this AS, so
+        # we don't yet know if a row for it exists; hence we have to upsert here.
+        await self.db_pool.simple_upsert(
+            table="application_services_state",
+            keyvalues={"as_id": service.id},
+            values={f"{stream_type}_stream_id": pos},
+            # no need to lock when emulating upsert: as_id is a unique key
+            lock=False,
+            desc="set_appservice_stream_type_pos",
         )
```
Comment (richvdh) [inline synapse/storage/databases/main/appservice.py:466]: I think this is actually fixing a previously-existing race. Since the row for `service.id` wasn't added until the transaction was successfully sent - which happened asynchronously, if at all - there was no guarantee that such a row actually existed at this point.

Of course, removing the code to add the row when the txn was sent meant that the row never got added at all, so the race became a consistent failure.

TL;DR: I assert this should have been an upsert all along.

File: synapse/storage/databases/main/appservice.py [hunk 2]
```diff
@@ -203,19 +203,29 @@ async def get_appservice_state(
...
             "application_services_state",
             {"as_id": service.id},
-            ["state"],
+            retcol="state",
             allow_none=True,
             desc="get_appservice_state",
         )
         if result:
-            return ApplicationServiceState(result.get("state"))
+            return ApplicationServiceState(result)
```
Comment (richvdh) [inline synapse/storage/databases/main/appservice.py:228]: this changes things so that if there is a row for the AS but it has `state IS NULL`, we return `None` rather than raising an exception.

And I *think* this is necessary because fixing the race in  `set_appservice_stream_type_pos` now means that we hit this code at a point where we have a row with `{stream_type}_stream_id` but not yet a `state`. Previously, there would have (incorrectly) been no row at all.

**Review by DMRobertson — CHANGES_REQUESTED** [review]
There's a merge conflict which needs resolving.

**Comment by richvdh** [comment]
oops. thanks, fixed.

File: synapse/storage/databases/main/appservice.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/appservice.py:228]: > this changes things so that if there is a row for the AS but it has `state IS NULL`, we return `None` rather than raising an exception.

It took me a second to understand that `ApplicationServiceState` is a enum, and so raises if given a `None`.

We appear to have changed this behaviour https://github.com/matrix-org/synapse/pull/11488, and previously we returned `None` if `state` was `None`.

**Review by erikjohnston — APPROVED** [review]
